### PR TITLE
Improve user-facing isomorphism and canonical labelling functions, especially with colourings

### PR DESCRIPTION
**Pull request by wilfwilson**
This PR adds support for isomorphism testing, and isomorphism finding, for *coloured* digraphs.

---

Previously in the Digraphs package... you could ask for the automorphism group of a coloured digraph without multiple edges (i.e. the subgroup of the automorphism of the digraph consisting of those elements that preserve the colouring), and you could ask for the canonical labelling of a coloured digraph without multiple edges.

I've now added similar support for coloured digraphs to the operations `IsIsomorphicDigraph` and `IsomorphismDigraphs`, for testing whether two coloured digraphs without multiple edges are isomorphic as coloured digraphs (i.e. whether there exists an isomorphism between the digraphs that respects the colourings of the vertices).

In addition to these new features, I've made minor improvements to the existing code, such as checking in `IsIsomorphicDigraph` whether the two digraphs are equal before going on to compute a canonical labelling.

The main change of this type is a new global function `DIGRAPHS_ValidateVertexPartition`. Creating this function allows me to replace a large amount of duplicated code which existed to validate whether the "colouring" given as an argument to `AutomorphismGroup` or `DigraphCanonicalLabelling` was indeed a valid colouring of the relevant digraph. This is now also used by `IsIsomorphicDigraph` and `IsomorphismDigraphs`.

Given a non-negative integer `n` and a homogeneous list `partition`, this `DIGRAPHS_ValidateVertexPartition` tests whether `partition` is a valid partition `[1 .. n]`. A valid partition of `[1 .. n]` is either:
1. a list of length `n` consisting of numbers in the range `[0 .. n]` (so that `partition[i]` is the colour of vertex `i`), or
2. a list of non-empty disjoint lists whose union is `[1 .. n]` (so that `partition[i]` is the list of those vertices with colour `i`).

If `partition` *is* a valid partition of `[1 .. n]` then it returns the partition in form 1 (this is the form that bliss likes); otherwise it returns `fail`.

This is both a more and less permissive definition of colouring than existed previously.

It's more permissive, since in the first form, I now allow the colour `0` to be used. This is because:

* bliss works fine with any colour in the range `[0 .. n]` (but not outside of this range), and
* in the Semigroups package, a method for `CanonicalBooleanMat` calculates the canonical labelling of a particular colouring digraph which uses colours 0 and 1. Since colour 0 is not currently allowed in Digraphs, this method uses `DIGRAPH_CANONICAL_LABELING_COLORS`. I think it would be better if this method used the operation `DigraphCanonicalLabelling` for a digraph and a colouring.

It's less permissive, since in the second form, I now require the list to consist of *disjoint* lists. This was currently allowed, but to my mind that was a mistake, as it doesn't make sense for a vertex to have two colours.

In time, this global function should also replace the validation of the colouring given to `HomomorphismDigraphsFinder`.

---

However, the main work on this PR is an improvement to documentation all of the functions `AutomorphismGroup`, `DigraphCanonicalLabelling`, `IsIsomorphicDigraph`, and `IsomorphismDigraphs` in the file `bliss.*`.

The previous documentation for these functions was either incorrect, incomplete, or confusing (this confusion prompted issue #19). It certainly left me scratching my head for a while. I've tried to make the documentation as clear, as thorough, and as helpful as possible.

---

As an aside, since in the `bliss.xml` documentation we talk a lot about colourings, where we simply mean a labelling, I now emphasise in the documentation for `DigraphColouring` and `ChromaticNumber` that these colourings are special in some sense -- they are *proper* colourings.

### CODE REVIEW

**Comment by wilfwilson** [comment]
Note: this PR makes no changes to anything at the C or C++ level, nor does any of the new code directly interface with any C functions. In particular, the new methods for `IsIsomorphicDigraph` and `IsomorphismDigraphs` call the already-existing `DigraphCanonicalLabelling` methods. Therefore this PR is entirely independent of any changes that we may or may not be making the software underpinning this package (see the discussion on issue #40).

File: doc/attr.xml [hunk 1]
```diff
@@ -13,14 +13,14 @@
   <Attr Name="ChromaticNumber" Arg="digraph"/>
   <Returns> A non-negative integer.</Returns>
   <Description>
-    A digraph coloring is a labeling of the vertices (using precisely <A>n</A>
```
Comment (james-d-mitchell) [inline doc/attr.xml]: This is weird, what is `n` here? This is not really related to your pull request, but since you are changing this man section anyway, why not remove the statement in brackets here? I guess this was copy pasted from another manual entry where the number of colours was specified by an argument `n`.

File: gap/bliss.gi [hunk 2]
```diff
@@ -12,141 +13,247 @@
...
-                    "the second arg must be a list of length ",
-                    DigraphNrVertices(graph), " of integers in [1 .. ",
-                    DigraphNrVertices(graph), "],");
-    fi;
-    colors := list;
+  n := DigraphNrVertices(digraph);
+  colours := DIGRAPHS_ValidateVertexPartition(n, list);
+  if colours = fail then
+    ErrorNoReturn("Digraphs: DigraphCanonicalLabelling: usage,\n",
+                  "the argument <list> does not define a valid colouring of ",
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: "Colouring" or "coloring"? There seem to be both in this file. I'm fine with either but we should probably be consistent.

File: gap/bliss.gi [hunk 3]
```diff
@@ -12,141 +13,247 @@
...
+                  "the vertices of\n<digraph>. Either <list> should be a list ",
+                  "of length <n> consisting of integers\nin the range ",
+                  "[0 .. <n>], or <list> should be a list of non-empty ",
+                  "disjoint\nlists whose union is the vertices of <digraph>,");
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: I think this error message could be more helpful. What is `n` for example? I think it would be better to add this error message to `DIGRAPHS_ValidateVertexPartition`and be more specific about what is wrong (as was previously attempted). I haven't yet looked at `DIGRAPHS_ValidateVertexPartition`, will comment more there.

File: gap/bliss.gi [hunk 4]
```diff
@@ -12,141 +13,247 @@
...
   fi;
-
-  return DIGRAPH_CANONICAL_LABELING_COLORS(graph, colors);
+  return DIGRAPH_CANONICAL_LABELING_COLORS(digraph, colours);
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: Shouldn't this be `LABELLING' for consistency?

File: gap/bliss.gi [hunk 5]
```diff
@@ -12,141 +13,247 @@
...
+
+#
+
+InstallMethod(IsIsomorphicDigraph, "for digraphs and homogeneous lists",
+[IsDigraph, IsDigraph, IsHomogeneousList, IsHomogeneousList],
+function(gr1, gr2, c1, c2)
+  local m, colour1, n, colour2, class_sizes, i;
+
+  if IsMultiDigraph(gr1) or IsMultiDigraph(gr2) then
+    ErrorNoReturn("Digraphs: IsIsomorphicDigraph: usage,\n",
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: Why is this not supported? On the face of it, isn't this just the same as the non-coloured case?

File: gap/bliss.gi [hunk 6]
```diff
@@ -12,141 +13,247 @@
...
+  fi;
+
+  class_sizes := ListWithIdenticalEntries(n, 0);
+  for i in DigraphVertices(gr1) do
+    class_sizes[colour1[i]] := class_sizes[colour1[i]] + 1;
+    class_sizes[colour2[i]] := class_sizes[colour2[i]] - 1;
+  od;
+  if not ForAll(class_sizes, x -> x = 0) then
+    return false;
+  elif gr1 = gr2 and colour1 = colour2 then
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: Do we have any evidence that equality checking is faster than checking isomorphism? I don't really have any preconceptions, except that finding a canonical labelling is usually very fast, so I'm just checking.

File: gap/bliss.gi [hunk 7]
```diff
@@ -12,141 +13,247 @@
...
-  if not IsIsomorphicDigraph(g1, g2) then
+  if not IsIsomorphicDigraph(gr1, gr2) then
     return fail;
   fi;
 
-  if IsMultiDigraph(g1) then
-    label1 := DigraphCanonicalLabelling(g1);
-    label2 := DigraphCanonicalLabelling(g2);
+  if IsMultiDigraph(gr1) then
+    if gr1 = gr2 then
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: Same comment as above, is this really better?

File: gap/bliss.gi [hunk 8]
```diff
@@ -12,141 +13,247 @@
...
+
+#
+
+InstallMethod(IsomorphismDigraphs, "for digraphs and homogeneous lists",
+[IsDigraph, IsDigraph, IsHomogeneousList, IsHomogeneousList],
+function(gr1, gr2, c1, c2)
+  local m, colour1, n, colour2, class_sizes, label1, label2, i;
+
+  if IsMultiDigraph(gr1) or IsMultiDigraph(gr2) then
+    ErrorNoReturn("Digraphs: IsomorphismDigraphs: usage,\n",
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: Again, why?

File: gap/bliss.gi [hunk 9]
```diff
@@ -12,141 +13,247 @@
...
+  class_sizes := ListWithIdenticalEntries(n, 0);
+  for i in DigraphVertices(gr1) do
+    class_sizes[colour1[i]] := class_sizes[colour1[i]] + 1;
+    class_sizes[colour2[i]] := class_sizes[colour2[i]] - 1;
+  od;
+  if not ForAll(class_sizes, x -> x = 0) then
+    return fail;
+  fi;
+
+  if gr1 = gr2 and colour1 = colour2 then
```
Comment (james-d-mitchell) [inline gap/bliss.gi]: `elif` here as above, please.

File: gap/utils.gi [hunk 10]
```diff
@@ -350,6 +350,64 @@ function(graph)
...
+# of the list [1 .. n]. A valid partition of [1 .. n] is either:
+#
+# 1. A list of length <n> consisting of numbers in the range [0 .. n]
+# 2. A list of non-empty disjoint lists whose union is [1 .. n]
+#
+# If <partition> is a valid partition of [1 .. n] then this global function
+# returns the partition, in form 1 (converting it to this form if necessary).
+# If <partition> is invalid, then the function returns <fail>
+
+InstallGlobalFunction(DIGRAPHS_ValidateVertexPartition,
```
Comment (james-d-mitchell) [inline gap/utils.gi]: Shouldn't this be in `bliss.gi`?

File: gap/utils.gi [hunk 11]
```diff
@@ -350,6 +350,64 @@ function(graph)
...
+function(n, partition)
+  local colours, vertices, i, vertex;
```
Comment (james-d-mitchell) [inline gap/utils.gi]: Why not take third argument which is a string containing the name of the function calling `DIGRAPHS_ValidateVertexPartition`? Then you could have specific error messages in this function rather than the same repeated error message in the functions calling this one.

File: gap/utils.gi [hunk 12]
```diff
@@ -350,6 +350,64 @@ function(graph)
...
+# Given a non-negative integer <n> and a homogeneous list <partition>,
+# this global function tests whether <partition> is a valid partition
+# of the list [1 .. n]. A valid partition of [1 .. n] is either:
+#
+# 1. A list of length <n> consisting of numbers in the range [0 .. n]
+# 2. A list of non-empty disjoint lists whose union is [1 .. n]
+#
+# If <partition> is a valid partition of [1 .. n] then this global function
+# returns the partition, in form 1 (converting it to this form if necessary).
+# If <partition> is invalid, then the function returns <fail>
```
Comment (james-d-mitchell) [inline gap/utils.gi]: Missing full stop

File: gap/utils.gi [hunk 13]
```diff
@@ -350,6 +350,64 @@ function(graph)
...
+    return fail;
+  fi;
+
+  # is <partition> a list of colours or a list of lists of vertices?
+
+  if IsPosInt(partition[1]) and Length(partition) = n then
+    colours := [];
+    for i in partition do
+      if not IsInt(i) or i < 0 or n < i then
+        return fail;
```
Comment (james-d-mitchell) [inline gap/utils.gi]: Error message

File: gap/utils.gi [hunk 14]
```diff
@@ -350,6 +350,64 @@ function(graph)
...
+                  "the first argument <n> must be a non-negative integer, and ",
+                  "the second argument\n<partition> must be a homogeneous ",
+                  "list,");
+  elif n = 0 then
+    if IsEmpty(partition) then
+      return partition;
+    fi;
+    return fail;
+  elif IsEmpty(partition) then
+    return fail;
```
Comment (james-d-mitchell) [inline gap/utils.gi]: error

File: doc/bliss.xml [hunk 15]
```diff
@@ -11,83 +12,58 @@
 <#GAPDoc Label="AutomorphismGroupDigraph">
 <ManSection>
   <Attr Name="AutomorphismGroup" Label="for a digraph" Arg="digraph"/>
-  <Oper Name="AutomorphismGroup" Label="for a digraph and a homogeneous list"
-    Arg="digraph, colors"/>
-  <Returns>A permutation group.</Returns>
+  <Returns>A permutation group, or a direct product of permutation
```
Comment (james-d-mitchell) [inline doc/bliss.xml]: Isn't a direct product of perm groups a perm group?

File: doc/bliss.xml [hunk 16]
```diff
@@ -11,83 +12,58 @@
...
+    groups.</Returns>
   <Description>
-    If <A>digraph</A> is a digraph without multiple edges, then this function
-    returns the automorphism group of <A>digraph</A>, as a group of
-    permutations on the vertices of <A>digraph</A>. <P/>
+    If <A>digraph</A> is a digraph, then this attribute returns the group of
```
Comment (james-d-mitchell) [inline doc/bliss.xml]: `returns` -> `stores` or `contains` maybe?

File: doc/bliss.xml [hunk 17]
```diff
@@ -11,83 +12,58 @@
...
+        If <A>digraph</A> is a multidigraph, then the automorphism group is
+        a group of permutations on the vertices and edges of <A>digraph</A>.
+        <P/>
+
+        For convenience, the group is returned as the direct product of the
+        group of automorphisms of the vertices of <A>digraph</A> with the
+        stabiliser of the vertices in the automorphism group of the edges. These
+        two groups can be accessed using the operation <Ref Oper="Projection"
+          Label="for a domain and a positive integer" BookName="ref"/>, with the
+        second argument being <C>1</C> or <C>2</C>, respectively. <P/>
```
Comment (james-d-mitchell) [inline doc/bliss.xml]: This is really good, can we say which projection corresponds to which action, and actually give an example? This would be most helpful.

File: doc/bliss.xml [hunk 18]
```diff
@@ -98,72 +74,143 @@ gap> Range(Projection(G, 2));
...
+
+      This operation computes the automorphism group of a coloured digraph.
+      Currently, it is only possible to compute the automorphism group of a
+      coloured digraph in the &Digraphs; package if the digraph has no multiple
+      edges; see <Ref Prop="IsMultiDigraph" />. <P/>
+
+      A coloured digraph can be specified by its underlying digraph
+      <A>digraph</A> and its colouring <A>colours</A>. Let <C>n</C> be the
+      number of vertices of <A>digraph</A>. The colouring <A>colours</A> may
+      have one of the following two forms: <P/>
```
Comment (james-d-mitchell) [inline doc/bliss.xml]: I think the `<P/>` here is superfluous.

File: doc/attr.xml (hunk 1 above)
Comment (wilfwilson) [inline doc/attr.xml]: Yes this must have been copied from the documentation for `DigraphColouring` when the documentation for `ChromaticNumber` was add. I'll reword it.

File: gap/bliss.gi (hunk 2 above)
Comment (wilfwilson) [inline gap/bliss.gi]: I favour colouring. I'll aim for consistency (including local variable names).

File: gap/bliss.gi (hunk 4 above)
Comment (wilfwilson) [inline gap/bliss.gi]: Yes, I'll make this change. The only reason I didn't change the names of the C functions is because the Semigroups package called one of these directly and I didn't want to have to fiddle with that. However I'll make a PR to Semigroups to sort this out.

File: gap/utils.gi (hunk 10 above)
Comment (wilfwilson) [inline gap/utils.gi]: I envisage `DIGRAPHS_ValidateVertexPartition` being used by the methods in `grahom.gi` as well, so I thought this would be the best place for it.

File: gap/bliss.gi (hunk 6 above)
Comment (wilfwilson) [inline gap/bliss.gi]: In a worst-case complexity sense, surely isomorphism testing is harder. For equality, it's a bit worse than `O(m + n)` since the out-neighbours of a vertex aren't required to be sorted. For each vertex of the two digraphs, you sort the respective out-neighbour list, and then compare, which doesn't seem that bad. It's implemented in `C` so it's fairly fast.

Here's some experimental evidence which I'd say shows that it is worth doing equality checking, although it's not dramatic.

```
gap> gr1 := CycleDigraph(50000);;
gap> gr2 := CycleDigraph(50000);;
gap> DigraphCanonicalLabelling(gr1);; 2 * time;
468
gap> gr1 = gr2;; time;
3
```

```
gap> gr1 := RandomDigraph(5000, 0.05);;
gap> gr2 := DigraphCopy(gr1);;
gap> DigraphCanonicalLabelling(gr1);; 2 * time;
198
gap> gr1 = gr2;; time;
4
```

```
gap> gr1 := RandomDigraph(5000, 0.2);;
gap> gr2 := DigraphCopy(gr1);;
gap> DigraphCanonicalLabelling(gr1);; 2 * time;
608
gap> gr1 = gr2;; time;
16
```

```
gap> gr1 := RandomDigraph(10000, 0.25);;
gap> gr2 := DigraphCopy(gr1);;
gap> DigraphCanonicalLabelling(gr1);; 2 * time;
3334
gap> gr1 = gr2;; time;
81
```

Comment (wilfwilson) [inline gap/bliss.gi]: (I do `2 * time;` for the canonical labelling since you've got to calculate it twice.)

File: doc/bliss.xml (hunk 15 above)
Comment (wilfwilson) [inline doc/bliss.xml]: Yes. `AutomorphismGroup` for a multidigraph previously had its own manual section, but I combined it with the section for a digraph without multiple edges (but without colours). The returns bit for the multidigraph section said "A direct product of permutation groups", so I just kept that bit.

File: doc/bliss.xml (hunk 17 above)
Comment (wilfwilson) [inline doc/bliss.xml]: You probably like it because you wrote it :) I'll do what I can to improve it.

**Comment by wilfwilson** [comment]
I will modify `DIGRAPHS_ValidateVertexPartition` as you describe, to display a more relevant error message, rather than returning `fail` and the calling function displaying a less-relevant error message.

The functions with colourings that I wrote do not support multidigraphs since the two-argument `DigraphCanonicalLabelling` does not support multidigraphs, it just returns `fail` (although this behaviour is undocumented in the released package). I'm not sure why coloured multidigraph canonical labelling was not added. I'll see if I can work out whether it is easy to add.

I'm happy to disallow the colour 0.

Another question: for a colouring of the form `[[1, 4], [3], [2, 5]]`, we disallow empty sublists (since then it's not a homogeneous list), and this implies that the colours used have to be precisely some range `[0 .. m]`. However, for a colouring of the form `[1, 3, 3, 1]`, we do not currently require that the used colours form a range, so for instance in this last example the colours used are `[1, 3]`. How do you think we should resolve this inconsistency? Either the first form allows empty lists, or the second doesn't allow colours to be skipped? Or we leave it as-is. I suggest we leave it as it is.

**Comment by wilfwilson** [comment]
I've answered my own question. This example uses the unique colour `1` for the first digraph, and the unique colour `2` for the second digraph. These aren't isomorphic as coloured digraphs since the colours don't match up! It's necessary for two coloured digraphs to have the same number of vertices of each particular colour in order to be isomorphic.

But the following code returns true rather than the (correct) false:
```
gap> a1 := Digraph([[2],[]]);;
gap> a2 := Digraph([[],[1]]);;
gap> q1 := DigraphCanonicalLabelling(a1, [1, 1]);;
gap> q2 := DigraphCanonicalLabelling(a2, [2, 2]);;
gap> OnDigraphs(a1, q1) = OnDigraphs(a2, q2);
true
```
(Note: this is hence a bug in the released version of the package).

Therefore a colouring of the form `[2, 2]` (not using the colour `1`) should be disallowed.

**Comment by wilfwilson** [comment]
I've revisted what I had written before. However, for consistency's sake, I still think it's right that a colouring of the form `[1, 3, 2, 3, 1]` shouldn't miss out colours.

**Comment by wilfwilson** [comment]
I think I've sorted all of this out now.

The main change I've made since your review was to add support for passing coloured multidigraphs to bliss. This mostly involved creating a function for constructing coloured multidigraphs for bliss, and a slightly different hook function. Coloured non-multidigraphs used one trick to encode their colours and directed edges, whilst non-coloured multidigraphs used another trick to encode their multiple edges. I just combined these things together.

That means we can ask about the automorphism group and canonical labelling (and hence isomorphism) of a coloured multidigraph. It seems to work fine.

**Comment by james-d-mitchell** [comment]
@wilfwilson I agree about not missing colours, I think this is the only sensible option.

File: gap/utils.gi (hunk 10 above)
Comment (james-d-mitchell) [inline gap/utils.gi]: Right but `utils.g*` only contains things relating to the administration of the package, and not its digraphs functionality. I suggest moving this to `bliss.g*`

File: doc/bliss.xml (hunk 15 above)
Comment (james-d-mitchell) [inline doc/bliss.xml]: Fair enough

File: doc/bliss.xml (hunk 17 above)
Comment (james-d-mitchell) [inline doc/bliss.xml]: Right :1st_place_medal:

File: doc/z-chap6.xml [hunk 19]
```diff
@@ -42,7 +45,7 @@ from} $E_a$ \emph{to} $E_b$. In this case we say that $E_a$ and $E_b$ are
     <#Include Label="DigraphEpimorphism">
     <#Include Label="EpimorphismsDigraphs">
     <#Include Label="GeneratorsOfEndomorphismMonoid">
-    <#Include Label="DigraphColoring">
+    <#Include Label="DigraphColouring">
```
Comment (james-d-mitchell) [inline doc/z-chap6.xml]: I think we'll need to keep `DigraphColoring` too for backwards compatibility.

File: gap/utils.gi [hunk 20]
```diff
@@ -350,6 +350,113 @@ function(graph)
...
+                      "the argument <partition> does not define a ",
+                      "colouring of the vertices [1 .. ", n, "],\nsince ",
+                      "it does not assign a colour to the vertex ", i, ",");
+      fi;
+      return colours;
+    fi;
+  fi;
+  ErrorNoReturn("Digraphs: ", method, ": usage,\n",
+                "the argument <partition> does not define a ",
+                "colouring of the vertices [1 .. ", n, "].\nTo do so, the ",
```
Comment (james-d-mitchell) [inline gap/utils.gi]: Remove the `To do so`

File: doc/z-chap6.xml (hunk 19 above)
Comment (wilfwilson) [inline doc/z-chap6.xml]: The synonym already exists in `grahom.gd`, and there are tests using both `DigraphColoring` and `DigraphColouring`. I just changed the name of the mansection's label.

**Comment by wilfwilson** [comment]
@james-d-mitchell All sorted.
